pkg/tsdb/azuremonitor/httpclient.go

```diff
@@ -8,34 +8,39 @@ import (
 	"github.com/grafana/grafana/pkg/tsdb/azuremonitor/aztokenprovider"
 )
 
-func httpClientProvider(route azRoute, model datasourceInfo, cfg *setting.Cfg) (*httpclient.Provider, error) {
-	var clientProvider *httpclient.Provider
+func getMiddlewares(route azRoute, model datasourceInfo, cfg *setting.Cfg) ([]httpclient.Middleware, error) {
+	middlewares := []httpclient.Middleware{}
 
 	if len(route.Scopes) > 0 {
 		tokenProvider, err := aztokenprovider.NewAzureAccessTokenProvider(cfg, model.Credentials)
 		if err != nil {
 			return nil, err
 		}
+		middlewares = append(middlewares, aztokenprovider.AuthMiddleware(tokenProvider, route.Scopes))
+	}
 
-		clientProvider = httpclient.NewProvider(httpclient.ProviderOptions{
-			Middlewares: []httpclient.Middleware{
-				aztokenprovider.AuthMiddleware(tokenProvider, route.Scopes),
-			},
+	if _, ok := model.DecryptedSecureJSONData["appInsightsApiKey"]; ok && (route.URL == azAppInsights.URL || route.URL == azChinaAppInsights.URL) {
+		// Inject API-Key for AppInsights
+		apiKeyMiddleware := httpclient.MiddlewareFunc(func(opts httpclient.Options, next http.RoundTripper) http.RoundTripper {
+			return httpclient.RoundTripperFunc(func(req *http.Request) (*http.Response, error) {
+				req.Header.Set("X-API-Key", model.DecryptedSecureJSONData["appInsightsApiKey"])
+				return next.RoundTrip(req)
+			})
 		})
-	} else {
-		clientProvider = httpclient.NewProvider()
+		middlewares = append(middlewares, apiKeyMiddleware)
 	}
 
-	return clientProvider, nil
+	return middlewares, nil
 }
 
-func newHTTPClient(route azRoute, model datasourceInfo, cfg *setting.Cfg) (*http.Client, error) {
-	model.HTTPCliOpts.Headers = route.Headers
-
-	clientProvider, err := httpClientProvider(route, model, cfg)
+func newHTTPClient(route azRoute, model datasourceInfo, cfg *setting.Cfg, clientProvider httpclient.Provider) (*http.Client, error) {
+	m, err := getMiddlewares(route, model, cfg)
 	if err != nil {
 		return nil, err
 	}
+	routeHTTPClient := model.HTTPCliOpts
+	routeHTTPClient.Headers = route.Headers
+	routeHTTPClient.Middlewares = m
```

Comment:
It doesn't look clean that you have to update the same instance of `model.HTTPCliOpts` several times for each datasource service

Reply:
You probably don't even need to use the `httpclient.Options` from `backend.DataSourceInstanceSettings` ([here](https://github.com/grafana/grafana/blob/1140824427d7e83b1c626560f2ed6799c89d049f/pkg/tsdb/azuremonitor/azuremonitor.go#L122)) because they contains generic HTTP configuration like basic auth, SigV4 etc ([here](https://github.com/grafana/grafana-plugin-sdk-go/blob/843e160c3b78b11a1db78e3a477f3fc3ec20e9c9/backend/common.go#L95)) which isn't used by Azure Monitor datasource since it has own custom configuration and authentication.

It doesn't create any practical issue, so I'm OK.

Reply:
right, not really needed